File: python/game_rules.py
from board import Board


def apply_game_of_life_rules(board: Board) -> None:
    n_side = board.cells.__len__()
    life_board: Board = Board.make_blank(n_side)
    death_board: Board = Board.make_blank(n_side)

    for i in range(n_side):
        for j in range(n_side):
            n_live_neighbors = board.get_n_live_neighbors(i, j)

            if board.cells[i][j] and n_live_neighbors != 2 and n_live_neighbors != 3:
                death_board.cells[i][j] = True
            if not board.cells[i][j] and n_live_neighbors == 3:
                life_board.cells[i][j] = True

    for i in range(n_side):
        for j in range(n_side):
            if death_board.cells[i][j]:
                board.cells[i][j] = False
            if life_board.cells[i][j]:
                board.cells[i][j] = True

File: rust/src/game_rules.rs
use crate::board::Board;

pub fn apply_game_of_life_rules(board: &mut Board) -> () {
    let (life_board, death_board) = get_life_death_boards(board);
    update_board(board, life_board, death_board);
}

fn update_board(board: &mut Board, life_board: Board, death_board: Board) -> () {
    let n_side = board.cells.len();
    for i in 0..n_side {
        for j in 0..n_side {
            if death_board.cells[i][j] {
                board.cells[i][j] = false;
            }
            if life_board.cells[i][j] {
                board.cells[i][j] = true;
            }
        }
    }
}

fn get_life_death_boards(board: &mut Board) -> (Board, Board) {
    let n_side = board.cells.len();

    let mut life_board = Board::new(n_side);
    let mut death_board = Board::new(n_side);

    for i in 0..n_side {
        for j in 0..n_side {
            let n_live_neighbors = board.get_n_live_neighbors(i, j);

            if board.cells[i][j] && n_live_neighbors != 2 && n_live_neighbors != 3 {
                death_board.cells[i][j] = true;
            }
            if !board.cells[i][j] && n_live_neighbors == 3 {
                life_board.cells[i][j] = true;
            }
        }
    }

    return (life_board, death_board);
}
